tc/core/compiler-inl.h

```diff
@@ -0,0 +1,101 @@
+  if (inputsInfo.size() != halideComponents.inputs.size()) {
+    throw lang::ErrorReport(halideComponents.getDef())
+        << "expected " << halideComponents.inputs.size() << " inputs but found "
+        << inputsInfo.size();
+  }
+  for (size_t i = 0; i < inputsInfo.size(); ++i) {
+    auto dltype = inputsInfo[i]->dtype;
+    auto htype = halideComponents.inputs[i].type();
+    // we have three type representations here: (1) halide Type (2) DLTensor
+    // type, and (3) the token representing the type in the frontend (e.g.
+    // TK_FLOAT) we need to translate to (3) to report user facing errors
+    auto dltype =
+        lang::TypeInfo(lang::TypeInfo::Code(dltype.code), dltype.bits)
+            .toScalarToken();
+    auto htype =
+        lang::TypeInfo(lang::TypeInfo::Code(htype.code()), htype.bits())
+            .toScalarToken();
+    if (dltype != htype) {
+      throw lang::ErrorReport(halideComponents.getDef().params()[i])
+          << "expected type " << lang::kindToString(htype) << " but found "
+          << lang::kindToString(dltype);
+    }
+    int hdim = halideComponents.inputs[i].dimensions();
+    int dldim = inputsInfo[i]->ndim;
+    if (dldim != hdim) {
+      throw lang::ErrorReport(halideComponents.getDef().params()[i])
+          << "expected a tensor with " << hdim << " dimensions but found "
+          << dldim << " dimensions.";
+    }
+  }
+}
+} // namespace
+template <typename Backend>
+std::unique_ptr<typename Backend::ExecutorType> compile(
+    const std::string& tc,
+    const std::string& entryPoint,
+    const std::vector<const DLConstTensor*>& inputs,
+    /* TODO: in the future also pass outputs for stride and alignment info */
+    const typename Backend::MappingOptionsType& options) {
+  auto parsedTcs = detail::parse(tc);
+  CHECK_EQ(1, parsedTcs.count(entryPoint))
+      << "attempting to access undefined function " << entryPoint;
+  return compile<Backend>(parsedTcs[entryPoint], inputs, options);
+}
+
+template <typename Backend>
+std::unique_ptr<typename Backend::ExecutorType> compile(
+    lang::TreeRef tcDefinition,
+    const std::vector<const DLConstTensor*>& inputs,
+    /* TODO: in the future also pass outputs for stride and alignment info */
+    const typename Backend::MappingOptionsType& options) {
+  using CompilationResultType = typename Backend::CompilationResultType;
+
+  auto inputsInfo = makeTensorInfoVector(inputs);
+  auto outputsInfo = detail::inferOutputTensorInfo(tcDefinition, inputs);
+  auto halideComponents =
+      tc2halide::translate(isl::with_exceptions::globalIslCtx(), tcDefinition);
+  checkInputsCompliant(halideComponents, inputs);
+
+  auto tcName = lang::Def(tcDefinition).name().name();
+  CompilationResultType compilationResult = Backend::compileWithTcMapper(
+      tcName,
+      halideComponents,
+      inputs,
+      /* TODO outputs, */
+      options);
+  return std::unique_ptr<typename Backend::ExecutorType>(
```

Comment:
when using this make_unique, please use tc:: namespace, btw. Otherwise it causes compilation errors in a project which uses c++14 and includes <memory>

Reply:
not using something custom here, we'll use the C++14 variant when we decide to pull the trigger on C++11-only